server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -175,8 +176,10 @@ private static long calculateRamUsage(long maxMapSize) {
 
     @Override
     public State reduce(RamAccounting ramAccounting, State state1, State state2) {
-        if (state1 instanceof TopKState t1 && state2 instanceof TopKState t2) {
-            return new TopKState(t1.itemsSketch.merge(t2.itemsSketch), t1.limit);
+        if (state1 instanceof TopKLongState t1 && state2 instanceof TopKLongState t2) {
+            return new TopKLongState(t1.sketch.merge(t2.sketch), t1.limit);
+        } else if (state1 instanceof TopKState t1 && state2 instanceof TopKState t2) {
+            return new TopKState(t1.sketch.merge(t2.sketch), t1.limit);
         } else if (state1 instanceof Empty) {
             return state2;
         } else {
```

Comment:
Seems like the `State` interface could use a `.merge()` and `.update` method?

Similar for the streaming and construction.

`State` could have a `.writeTo()` to encapsulate the data writing logic and might be worth adding a `id()` property or so to keep the type distinction also central in the `State` class. 

Currently it looks like a bit of a odd mix between inheritance polymorphism and instanceof dispatching.
Ideally all the differences would be encapsulated within `State`, and the remainder of `TopKAggregation` wouldn't be aware of the differences.